File: src/memory/docs_lint.py
"""Link and anchor integrity checker for the project's markdown documentation.

Scans every ``*.md`` file in the repository for relative file links and
in-document/cross-document anchor links, verifying that:

- every relative file target actually exists on disk;
- every ``#anchor`` resolves to a real heading in the target file, using the
  same slug algorithm GitHub applies to render heading ids (lowercase,
  ``[a-z0-9 _-]`` kept, everything else stripped, spaces become hyphens,
  duplicate headings get a numeric suffix).

Deliberately network-free and deterministic: it never checks external
(``http``/``https``) links, only repo-relative ones. External-link liveness
is flaky and rate-limited, and is a non-goal (see the "Testing" section of
docs/process/engineering-principles.md).

Also checks a second, unrelated invariant that happens to share this CI job:
``docs/project/roadmap/**/index.md`` heading uniqueness. Two roadmap packages
must never claim the same ``# <code> — title`` heading code -- that is the
static, pre-runtime guard for the "one code -> one package" invariant Ariad's
Expand depends on (``memory.builder.story_paths.resolve_story_directory``),
closing the class of defect first seen as CR048 (DS6.TS5) and recurring as
CV22.DS7 (see ``handoff-ariad-fix.MD`` / ``plan-ariad-fix.md``).

Used by ``scripts/check_doc_links.py`` (CLI) and the ``docs`` CI workflow.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path

from memory.builder.story_paths import find_duplicate_roadmap_headings

_LINK_RE = re.compile(r"\[[^\]]*\]\(([^)]+)\)")
_HEADING_RE = re.compile(r"^#{1,6}\s+(.+?)\s*$", re.MULTILINE)
_FENCE_RE = re.compile(r"```.*?```", re.DOTALL)
_SLUG_STRIP_RE = re.compile(r"[^a-z0-9 _\-]")

_EXCLUDED_DIR_NAMES = {"node_modules", ".venv", "__pycache__"}

# roadmap/templates/*.md link to a bare `index.md` as a breadcrumb meant to
# resolve once the template is copied into a real story folder -- templates/
# itself has no index.md, by design. Not a broken link.
_TEMPLATE_DIR_MARKER = "/roadmap/templates/"


@dataclass(frozen=True)
class BrokenLink:
    """One link or anchor that does not resolve."""

    source_file: str
    line: int
    target: str
    reason: str


@dataclass(frozen=True)
class DuplicateRoadmapHeading:
    """Two or more roadmap ``index.md`` files whose heading claims the same code."""

    code: str
    paths: tuple[str, ...]


def slugify(heading: str) -> str:
    """Reproduce GitHub's heading-to-anchor-id algorithm.

    Lowercase, strip everything except letters/digits/spaces/underscores/
    hyphens (existing hyphens and underscores are preserved -- a heading
    like ``MEMORY_LOG_LLM_CALLS`` slugs to ``memory_log_llm_calls``, not
    ``memory-log-llm-calls``), then replace spaces with hyphens. Repeated
    hyphens are never collapsed, matching GitHub's real behavior.
    """
    lowered = heading.lower()
    stripped = _SLUG_STRIP_RE.sub("", lowered)
    return stripped.replace(" ", "-")


def strip_fenced_code(text: str) -> str:
    """Blank out fenced code blocks so link-like text inside ``` ``` is
    never scanned, while preserving line count exactly -- so any problem
    reported after a fence still points at the real source line.
    """

    def _blank(match: re.Match[str]) -> str:
        return "\n" * match.group(0).count("\n")

    return _FENCE_RE.sub(_blank, text)


def extract_headings(text: str) -> list[str]:
    """Return every Markdown heading's text, in document order."""
    return [m.group(1) for m in _HEADING_RE.finditer(text)]


def compute_anchors(text: str) -> set[str]:
    """Return every valid anchor id reachable in this document.

    Duplicate heading text is suffixed ``-1``, ``-2``, ... in encounter
    order, matching GitHub's disambiguation for repeated headings.
    """
    anchors: set[str] = set()
    seen_counts: dict[str, int] = {}
    for heading in extract_headings(text):
        base = slugify(heading)
        count = seen_counts.get(base, -1) + 1
        seen_counts[base] = count
        anchors.add(base if count == 0 else f"{base}-{count}")
    return anchors


def _is_template_placeholder(source_file: Path, target_path: str) -> bool:
    return _TEMPLATE_DIR_MARKER in source_file.as_posix() and target_path == "index.md"


def _iter_markdown_files(repo_root: Path) -> list[Path]:
    files = [
        path
        for path in repo_root.rglob("*.md")
        if not any(part in _EXCLUDED_DIR_NAMES for part in path.parts)
    ]
    return sorted(files)


def _line_of(text: str, offset: int) -> int:
    return text.count("\n", 0, offset) + 1


def check_file(source_file: Path, repo_root: Path) -> list[BrokenLink]:
    """Check every relative/anchor link in one markdown file."""
    raw = source_file.read_text(encoding="utf-8")
    scanned = strip_fenced_code(raw)
    own_anchors: set[str] | None = None
    problems: list[BrokenLink] = []
    rel_source = source_file.relative_to(repo_root).as_posix()

    for match in _LINK_RE.finditer(scanned):
        link = match.group(1).strip()
        if link.startswith(("http://", "https://", "mailto:")):
            continue
        line = _line_of(scanned, match.start())

        if link.startswith("#"):
            if own_anchors is None:
                own_anchors = compute_anchors(raw)
            if link[1:] not in own_anchors:
                problems.append(BrokenLink(rel_source, line, link, "anchor not found in this file"))
            continue

        path_part, _, anchor_part = link.partition("#")
        if _is_template_placeholder(source_file, path_part):
            continue

        target = Path(os.path.normpath(source_file.parent / path_part))
        if not target.exists():
            problems.append(BrokenLink(rel_source, line, link, "target file does not exist"))
            continue

        if anchor_part and target.suffix == ".md":
            target_anchors = compute_anchors(target.read_text(encoding="utf-8"))
            if anchor_part not in target_anchors:
                problems.append(
                    BrokenLink(rel_source, line, link, "anchor not found in target file")
                )

    return problems


def check_repo(repo_root: Path) -> list[BrokenLink]:
    """Check every markdown file in the repository. Deterministic order."""
    problems: list[BrokenLink] = []
    for md_file in _iter_markdown_files(repo_root):
        problems.extend(check_file(md_file, repo_root))
    return problems


def check_roadmap_duplicate_headings(repo_root: Path) -> list[DuplicateRoadmapHeading]:
    """Check ``docs/project/roadmap`` for two ``index.md`` files sharing a heading code.

    Deterministic, sorted order (by code, then by path) for stable CI output.
    Returns an empty list when the roadmap directory does not exist or every
    code is unique. Resolves ``repo_root`` before computing relative paths --
    ``find_duplicate_roadmap_headings`` returns already-resolved directories
    internally, so an unresolved ``repo_root`` (e.g. containing a symlink,
    ``..``, or a relative path) would otherwise fail ``Path.relative_to``.
    """
    resolved_root = repo_root.resolve()
    duplicates = find_duplicate_roadmap_headings(repo_root)
    return [
        DuplicateRoadmapHeading(
            code=code,
            paths=tuple(
                sorted(str((path / "index.md").relative_to(resolved_root)) for path in paths)
            ),
        )
        for code, paths in sorted(duplicates.items())
    ]

File: ts/src/docs/docsLint.ts
// Docs link/anchor checker — the TypeScript port of `src/memory/docs_lint.py`
// (CV22.DS10.TS5, slice B).
//
// The Zero Python gate assigns `scripts/check_doc_links.py` to this story. The
// script is a 57-line wrapper; the subject is this module's Python original
// (221 lines) plus the duplicate-heading guard, which TypeScript already owns
// through `findDuplicateRoadmapHeadings` (DS7.US8). So this port is the
// link/anchor half, and the roadmap half is a thin adapter over existing code.
//
// Every behavior here is pinned by `ts/test/scripts/docsLint.test.ts`, which
// carries all 26 cases of `tests/unit/memory/test_docs_lint.py`. That test is
// the reason this file exists in the shape it does: each case records a
// failure mode the checker actually caught, or once missed, during the
// repo-wide docs audit. The Python self-test dies with its suite; the
// assertions do not.

import {
  type Dirent,
  existsSync,
  readdirSync,
  readFileSync,
  realpathSync,
  statSync,
} from "node:fs";
import { dirname, join, normalize, relative, resolve, sep } from "node:path";

import { roadmapPaths } from "#builder/pullCandidates.ts";
import { findDuplicateRoadmapHeadings } from "#builder/storyPaths.ts";

const LINK_RE = /\[[^\]]*\]\(([^)]+)\)/g;
const HEADING_RE = /^#{1,6}[ \t]+(.+?)[ \t]*$/gm;
const FENCE_RE = /```[\s\S]*?```/g;
const SLUG_STRIP_RE = /[^a-z0-9 _-]/g;

const EXCLUDED_DIR_NAMES = new Set(["node_modules", ".venv", "__pycache__", ".git"]);

// Parity fixture trees (`ts/test/fixtures/**`) are test INPUT, not
// documentation. They carry authored-Markdown edge cases an oracle is graded
// on -- dangling links, duplicate heading codes, malformed tables -- so
// linting them as docs would force the fixtures to stop being fixtures.
// Narrow on purpose: a directory named `fixtures` is skipped only beneath a
// `test`/`tests` directory, so a real `docs/**/fixtures/` tree is still
// checked.
const TEST_DIR_NAMES = new Set(["test", "tests"]);
const FIXTURE_DIR_NAME = "fixtures";

// `roadmap/templates/*.md` link to a bare `index.md` breadcrumb meant to
// resolve once the template is copied into a real story folder; `templates/`
// itself has no `index.md`, by design. Not a broken link.
const TEMPLATE_DIR_MARKER = "/roadmap/templates/";

export interface BrokenLink {
  readonly sourceFile: string;
  readonly line: number;
  readonly target: string;
  readonly reason: string;
}

export interface DuplicateRoadmapHeading {
  readonly code: string;
  readonly paths: readonly string[];
}

/**
 * Reproduce GitHub's heading-to-anchor algorithm.
 *
 * Lowercase, strip everything except letters/digits/spaces/underscores/
 * hyphens (existing hyphens and underscores are PRESERVED -- `MEMORY_ENV`
 * slugs to `memory_env`, not `memory-env`), then spaces become hyphens.
 * Repeated hyphens are never collapsed, matching GitHub: an em dash is
 * stripped outright and leaves its surrounding spaces behind, which is why
 * `D1 — Local-first` slugs to `d1--local-first`.
 */
export function slugify(heading: string): string {
  return heading.toLowerCase().replace(SLUG_STRIP_RE, "").replace(/ /g, "-");
}

/**
 * Blank out fenced code blocks so link-like text inside ``` ``` is never
 * scanned, preserving the line count exactly -- a problem reported after a
 * fence must still point at its real source line.
 */
export function stripFencedCode(text: string): string {
  return text.replace(FENCE_RE, (match) => "\n".repeat((match.match(/\n/g) ?? []).length));
}

/** Every Markdown heading's text, in document order. */
export function extractHeadings(text: string): string[] {
  const headings: string[] = [];
  for (const match of text.matchAll(HEADING_RE)) headings.push(match[1] as string);
  return headings;
}

/**
 * Every anchor id reachable in this document. Duplicate heading text is
 * suffixed `-1`, `-2`, ... in encounter order, as GitHub disambiguates.
 */
export function computeAnchors(text: string): Set<string> {
  const anchors = new Set<string>();
  const seen = new Map<string, number>();
  for (const heading of extractHeadings(text)) {
    const base = slugify(heading);
    const count = (seen.get(base) ?? -1) + 1;
    seen.set(base, count);
    anchors.add(count === 0 ? base : `${base}-${count}`);
  }
  return anchors;
}

function toPosix(path: string): string {
  return sep === "/" ? path : path.split(sep).join("/");
}

function isTemplatePlaceholder(sourceFile: string, targetPath: string): boolean {
  return toPosix(sourceFile).includes(TEMPLATE_DIR_MARKER) && targetPath === "index.md";
}

/** True for a `fixtures` directory nested under a `test`/`tests` one. */
function isTestFixturePath(path: string): boolean {
  const parts = toPosix(path).split("/");
  return parts.some(
    (part, index) =>
      part === FIXTURE_DIR_NAME && parts.slice(0, index).some((p) => TEST_DIR_NAMES.has(p)),
  );
}

function lineOf(text: string, offset: number): number {
  let line = 1;
  for (let i = 0; i < offset; i += 1) if (text[i] === "\n") line += 1;
  return line;
}

/** Every markdown file in the tree, in deterministic (sorted) order. */
function iterMarkdownFiles(repoRoot: string): string[] {
  const found: string[] = [];
  const walk = (dir: string): void => {
    let entries: Dirent[];
    try {
      entries = readdirSync(dir, { withFileTypes: true, encoding: "utf8" });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        if (EXCLUDED_DIR_NAMES.has(entry.name)) continue;
        walk(full);
      } else if (entry.isFile() && entry.name.endsWith(".md")) {
        if (!isTestFixturePath(full)) found.push(full);
      }
    }
  };
  walk(repoRoot);
  return found.sort();
}

/** Check every relative/anchor link in one markdown file. */
export function checkFile(sourceFile: string, repoRoot: string): BrokenLink[] {
  const raw = readFileSync(sourceFile, "utf8");
  const scanned = stripFencedCode(raw);
  const relSource = toPosix(relative(repoRoot, sourceFile));
  const problems: BrokenLink[] = [];
  let ownAnchors: Set<string> | null = null;

  for (const match of scanned.matchAll(LINK_RE)) {
    const link = (match[1] as string).trim();
    if (/^(https?:\/\/|mailto:)/.test(link)) continue;
    const line = lineOf(scanned, match.index ?? 0);

    if (link.startsWith("#")) {
      ownAnchors ??= computeAnchors(raw);
      if (!ownAnchors.has(link.slice(1))) {
        problems.push({
          sourceFile: relSource,
          line,
          target: link,
          reason: "anchor not found in this file",
        });
      }
      continue;
    }

    const hashIndex = link.indexOf("#");
    const pathPart = hashIndex === -1 ? link : link.slice(0, hashIndex);
    const anchorPart = hashIndex === -1 ? "" : link.slice(hashIndex + 1);
    if (isTemplatePlaceholder(sourceFile, pathPart)) continue;

    const target = normalize(join(dirname(sourceFile), pathPart));
    if (!existsSync(target)) {
      problems.push({
        sourceFile: relSource,
        line,
        target: link,
        reason: "target file does not exist",
      });
      continue;
    }

    if (anchorPart && target.endsWith(".md")) {
      if (!computeAnchors(readFileSync(target, "utf8")).has(anchorPart)) {
        problems.push({
          sourceFile: relSource,
          line,
          target: link,
          reason: "anchor not found in target file",
        });
      }
    }
  }

  return problems;
}

/** Check every markdown file in the repository. Deterministic order. */
export function checkRepo(repoRoot: string): BrokenLink[] {
  const problems: BrokenLink[] = [];
  for (const file of iterMarkdownFiles(repoRoot)) problems.push(...checkFile(file, repoRoot));
  return problems;
}

/**
 * Two or more roadmap `index.md` files whose heading claims the same code.
 *
 * Sorted by code, then by path, for stable CI output.
 *
 * Two resolution hazards, both real, both pinned by the self-test:
 *
 * 1. **The helper's contract is not Python's.** `find_duplicate_roadmap_headings`
 *    returns ABSOLUTE directories; the TypeScript `findDuplicateRoadmapHeadings`
 *    returns paths RELATIVE TO THE ROADMAP ROOT (`cv2-ds1-a`), because that is
 *    what `resolveStoryDirectory` needed when DS7.US8 ported it. Same name,
 *    different contract -- so this adapter joins them onto the roadmap root
 *    rather than treating them as absolute. Assuming parity of a helper whose
 *    name matched would have produced `../../../..`-prefixed garbage in CI
 *    output.
 * 2. **Both ends must be realpath'd or neither.** Python resolves the repo root
 *    and the roadmap root, so a root holding `..` or (on macOS) a
 *    `/var` -> `/private/var` symlink still yields a repo-relative path.
 *    Resolving only one end reintroduces exactly the defect the Python test
 *    says was caught by manual smoke-testing.
 */
export function checkRoadmapDuplicateHeadings(repoRoot: string): DuplicateRoadmapHeading[] {
  const realpathOrSelf = (path: string): string => {
    try {
      return statSync(path).isDirectory() ? realpathSync(path) : path;
    } catch {
      return path;
    }
  };

  const resolvedRoot = realpathOrSelf(resolve(repoRoot));
  const resolvedRoadmapRoot = realpathOrSelf(resolve(roadmapPaths(repoRoot).roadmapRoot));

  const duplicates = findDuplicateRoadmapHeadings(repoRoot);
  return [...duplicates.entries()]
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([code, directories]) => ({
      code,
      paths: directories
        .map((directory) =>
          toPosix(relative(resolvedRoot, join(resolvedRoadmapRoot, directory, "index.md"))),
        )
        .sort(),
    }));
}
